crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1804,70 +1807,145 @@ impl<'db> ClassType<'db> {
+    fn node(self, db: &'db dyn Db) -> &'db ast::StmtClassDef {
+        match self.definition(db).kind(db) {
+            DefinitionKind::Class(class_stmt_node) => class_stmt_node,
+            _ => unreachable!("Class type definition should always have DefinitionKind::Class"),
+        }
+    }
+
+    /// Attempt to resolve the [method resolution order] ("MRO") for this class.
+    /// If the MRO is unresolvable, return an error indicating why the class's MRO
+    /// cannot be accurately determined.
+    ///
+    /// The MRO is the tuple of classes that can be retrieved as the `__mro__`
+    /// attribute on a class at runtime.
+    ///
+    /// [method resolution order]: https://docs.python.org/3/glossary.html#term-method-resolution-order
+    #[salsa::tracked(return_ref)]
+    fn try_mro(self, db: &'db dyn Db) -> Result<Mro<'db>, MroError<'db>> {
+        Mro::of_class(db, self)
+    }
+
+    /// Iterate over the [method resolution order] ("MRO") of the class.
+    ///
+    /// If the MRO could not be accurately resolved, this method falls back to
+    /// an MRO that has the class directly inheriting from `Unknown`. Use
+    /// [`ClassType::try_mro`] if you need to distinguish between the success and failure
+    /// cases rather than simply iterating over the inferred resolution order for the class.
+    ///
+    /// [method resolution order]: https://docs.python.org/3/glossary.html#term-method-resolution-order
+    fn mro(self, db: &'db dyn Db) -> impl Iterator<Item = ClassBase<'db>> {
+        // We avoid materialising the *full* MRO unless it is actually necessary:
+        //
+        // - Materialising the full MRO is expensive
+        // - We need to do it for every class in the code that we're checking, as we need to make sure
+        //   that there are no class definitions in the code we're checking that would cause an
+        //   exception to be raised at runtime. But the same does *not* apply for every class
+        //   in third-party and stdlib dependencies: we never emit diagnostics about non-first-party code.
+        // - However, we *do* need to resolve attribute accesses on classes/instances from
+        //   third-party and stdlib dependencies. That requires iterating over the MRO of third-party/stdlib
+        //   classes, but not necessarily the *whole* MRO: often just the first element is enough.
+        //   Luckily we know that for any class `X`, the first element of `X`'s MRO will always be `X` itself.
+        //   We can therefore avoid resolving the full MRO for many third-party/stdlib classes while still
+        //   being faithful to the runtime semantics.
+        std::iter::once(ClassBase::Class(self)).chain(
+            [()].into_iter()
+                .flat_map(move |()| match self.try_mro(db) {
+                    Ok(mro) => mro.iter().copied(),
+                    Err(MroError { fallback_mro, .. }) => fallback_mro.iter().copied(),
+                })
+                .skip(1),
+        )
```

Comment:
Nit: It wasn't clear to me why the `[()]` was necessary and naively removed it to see if it still compiles.... and it does. But what I understand is that doing so removes the performance optimization. 

I suggest moving this code into a custom `Iterator` implementation to make the lazy lookup more explicit.